[#36] 특정 사용자에 대한 전체 추천글 보기 기능 구현

msugo1 opened the pull request:
## 📖 Controller 계층 관련

### 📑 RecommendationController

- 특정 사용자에 대한 전체 추천글 보기 요청을 매핑할 `getRecommendationsAboutTarget` 메소드 추가

- cursor 방식의 페이징 채택으로 cursor 값도 함께 서비스 계층으로 전달해준다.

## 📖 Service 계층 관련

### 📑 RecommendationService

- 특정 사용자에 대한 추천글 보기에 대한 요청을 Mapper 계층으로 전달할 `getRecommendationsAboutTarget` 메소드 추가

- 해당 사용자가 존재하지 않는 경우 `InvalidRequestException` 예외 발생

### 📑 UserService

- `특정 사용자가 이미 탈퇴했는지에 대한 여부를 판별`할 책임을 가진 `isUserActivated` 메소드 추가

### 📑 RecommendationService

- UserService를 주입 받아 위의 메소드를 사용해 특정 사용자에 대한 추천글 들을 받아오기 전 해당 사용자가 존재하는지 체크하기 위한 로직 추가

### 📑 BlockUserService/FriendService

- 해당 클래스에 존재하는 Validator 메소드에 넘겨줬던 기존 파라미터 값을 유저가 활성화 되어있는지의 여부를 포함한`isUserActivated`의 결과 값으로 변경

## 📖 Mapper 계층 관련

### 📑 RecommendationMapper

- cursor 위치부터주입된 pageSize 값 까지의 개수만큼 DB로 부터 추천글을 가져올 `getRecommendationsListAboutTarget` 메소드 추가

### 📑 RecommendationMapper.xml

- 해당 로직을 처리할 쿼리문 작성

- 유저 프로필에 출력할 추천글의 개수를 10개에서 5개로 수정

## 📖 Model 관련

### 📑 Recommendations

- DB에서 가져온 추천글 목록을 담아서 리턴하는데 사용할 객체를 만들 클래스

- 변수로는 id(int), from(해당 사용자 아이디, string), content(추천글 내용, string) 그리고 writtenAt(추천글 작성 날짜, timestamp)

### 📑 RecommendationForProfile

- 클래스 이름 변경

- 유저 프로필에 추천글 목록 출력시 추천글이 작성된 날짜 또한 포함될 수 있도록 멤버 변수 추가

### 📑 UserMapper

- 해당 사용자의 계정이 활성화된 상태인지 DB에서 확인해 줄 메소드 `isUserActivated` 추가

* UserMapper.xml

- 위의 내용을 처리할 쿼리문 추가

## 📖 테스트 관련

### 📑 RecommendationServiceTest

- 위의 로직을 검증할 테스트 작성

- 포스트맨을 활용해 전반적으로 테스트

Commit:
[#36] 특정 사용자에 대한 전체 추천글 보기 기능 구현

Controller 계층 관련

* RecommendationController

- 특정 사용자에 대한 전체 추천글 보기 요청을 매핑할 `getRecommendationsListAboutTarget` 메소드 추가

- cursor 방식의 페이징 채택으로 cursor 값도 함께 서비스 계층으로
  전달해준다.

Service 계층 관련

* RecommendationService

- 특정 사용자에 대한 추천글 보기에 대한 요청을 Mapper 계층으로 전달할 `getRecommendationsListAboutTarget` 메소드 추가

- 해당 사용자가 존재하지 않는 경우 `InvalidRequestException` 예외 발생

Mapper 계층 관련

* RecommendationMapper

- cursor 위치부터주입된 pageSize 값 까지의 개수만큼 DB로 부터 추천글을
  가져올 `getRecommendationsListAboutTarget` 메소드 추가

* RecommendationMapper.xml

- 해당 로직을 처리할 쿼리문 작성

- 유저 프로필에 출력할 추천글의 개수를 10개에서 5개로 수정

Model 관련

* RecommendationList

- DB에서 가져온 추천글 목록을 담아서 리턴하는데 사용할 객체를 만들
  클래스

- 변수로는 id(int), from(해당 사용자 아이디, string), content(추천글
  내용, string) 그리고 writtenAt(추천글 작성 날짜, timestamp)가 존재

* RecommendationForProfile

- 클래스 이름 변경

- 유저 프로필에 추천글 목록 출력시 추천글이 작성된 날짜 또한 포함될 수
  있도록 멤버 변수 추가

Commit:
[#36] 기능 구현 수정

- 포스트맨을 활용한 테스트 밑 예외를 발생시키는 라인 수정

Commit:
[#36] 수정 및 필요 기능 추가 업데이트

Service 계층 관련

* UserService

- `특정 사용자가 이미 탈퇴했는지에 대한 여부를 판별`할 책임을 가진 `isUserActivated` 메소드
  추가

* RecommendationService

- UserService를 주입 받아 위의 메소드를 사용해 특정 사용자에 대한 추천글 들을 받아오기 전 해당 사용자가 존재하는지체크하기 위한 로직 추가

* BlockUserService/FriendService

- 해당 클래스에 존재하는 Validator 메소드에 넘겨줬던 기존 파라미터 값을
  `isUserActivated`의 결과로 변경

Mapper 계층 관련

* UserMapper

- 해당 사용자의 계정이 활성화된 상태인지 DB에서 확인해 줄 메소드
  `isUserActivated` 추가

* UserMapper.xml

- 위의 내용을 처리할 쿼리문 추가

테스트 관련

* RecommendationServiceTest

- 위의 로직을 검증할 테스트 작성

- 포스트맨을 활용해 전반적으로 테스트

msugo1 commented:
#### `isUserActivated` 메소드 추가 이유

- 기존 `isUserIdExist` 를 활용해서 특정 사용자의 존재 여부를 체크했을 시, 사용자의 활성화 여부인 `isDeactivated` 조건을 체크하지 않았기 때문에 비활성화 된 유저의 경우에도 `사용자의 존재 유무 체크`를 통과함으로써 서비스 이용은 불가능하지만 해당 사용자에 대한 ` 등록된 추천글` 계속 조회가 가능하다는 문제점이 있었습니다.

- 이 부분은 서비스 운영 정책에 따라 다르겠지만, 논리 삭제를 활용해 일정 기간 동안 사용자의 계정관련 정보를 보관해 두는 경우도 많다고 들었습니다. 일차적으로는 `isUserIdExist`의 쿼리문에 `isDeactivated = 'N'` 조건을 추가해 문제를 해결하고자 했으나 그렇게 되면 사용자의 정보가  완전히 삭제되기 전, `비활성화기간 동안`에도 다시 해당 `비활성화된 아이디로 중복가입이 가능`했습니다. 

- 중복 가입여부 자체는 큰 문제가 아닐 수 있겠으나,  `DB 상에는 해당 유저가 다시 존재`하게 되어, 해당 ID에 등록된 기존 정보들이 다시 이용가능해지는 이슈가 발생했습니다. 프로필 내부에는 보통 공개적인 정보가 들어가기 때문에 크게 문제가 될일은 없으나 차후에 `메시지 기능`을 추가했을 때도 비슷한 문제가 발생하면 사적인 정보들과 관련해서 더 큰 이슈가 발생할 수 있으리라 생각했습니다. 

- 그래서 고민하다가 `아이디의 중복 여부`와 `사용자의 계정 비활성화 여부`의 체크를 담당하는 책임을 별도의 메소드로 분리하여 비활성화 후에도 정보가 남아있는 경우 `isUserIdExist` 메소드를 활용해 중복여부를 체크하여  `같은 아이디로의 중복 가입을 막고` 사용자의 계정 자체가 존재하는지(활성화)의 여부는 `isUserDeactivated`를 통해 체크함으로써 이 문제를 해결해보고자 하였습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/model/recommendation/RecommendationForProfile.java:
```diff
@@ -0,0 +1,19 @@
+    private String content;
+
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yy-MM-dd")
+    private Timestamp writtenAt;
```
여기서는 `Timestamp`를 사용하신 이유가 있으실까요? `LocalDate`가 더 좋아보입니다~

msugo1 replied:
추천글을 남길 때 남기는 시점을 등록하기 위해서 RDBMS에서 `NOW()`를 활용하는데 이때 타입이 `TimeStamp`여서 그렇게 지정해줬었습니다. `LocalDate`를 사용할 수 있는지는 몰랐네요. `LocalDate`로 변경하도록 하겠습니다.

Commit:
[#36] 피드백 반영

- 추천 글의 내용을 담은  객체를 생성하는 클래스의 인스턴스 멤버 변수 중
  `TimeStamp`를 `LocalDate`로 변경

Commit:
[#36] 코드 및 쿼리 수정

* 조회하려난 대상의 차단 목록에 등록되어 있는 경우 제대로 된 조회가 되지
  않도록 코드 변경 및 쿼리 변경

* 이에 따른 테스트 코드 수정

* `isUserActivated` 메소드 도입으로 코드가 변경되었으나 해당 내역이
  반영되지 않았던 테스트 메소드 수정

Commit:
[#36] LocalDate 직/역직렬화 시 오류 해결

- Jackson 라이브러리는 Date 관련 타입을 직렬화/역직렬화를 처리하지 못함

- ` @JsonSerialize(using = LocalDateSerializer.class), @JsonDeserialize(using = LocalDateDeserializer.class)`을 추가하여 일차적인 문제 해결

f-lab-dev commented on src/main/java/me/soo/helloworld/model/recommendation/RecommendationForProfile.java:
```diff
@@ -14,6 +17,7 @@
 
-    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yy-MM-dd")
+    @JsonSerialize(using = LocalDateSerializer.class)
+    @JsonDeserialize(using = LocalDateDeserializer.class)
```
이것도 `ObjectMapper`에 옵션을 줄 수 있을것 같아요~

Commit:
[#36] 피드백 반영

* RecommendationForProfile

- 기존의 어노테이션 JsonSerializer, Deserializer을 제거

* RedisCacheConfig

- Redis 캐시 저장소에 저장할 때 사용할 objectMapper 정의

- 타입이 제대로 인식되지 않아 UserProfile 클래스가 LinkedHashMap으로
  인식되어 역직렬화가 제대로 되지 않는 문제 발생

- 이를 해결하기 위해서 맞는 타입이 없는 경우 설정에 따라 타입을
  전달해주는 `DefaultTyping` 활성화 및 설정 추가

- objectMapper를 하나 더 정의해서 레디스 저장소에서 가져올 때와 캐시
  저장이 아닌 일반적인 직렬화/역직렬화 시에는 해당 objectMapper가 사용될
  수 있도록 하여 전달된 타입마저 표시되는 문제 해결

- ParameterNamesModule을 두 번째 objectMapper에 추가로 등록해 줌으로써
  역직렬화 시 기본 생성자나 어노테이션의 도움 없이도  없어도 AllArgsConstructor 내부의 파라미터 이름을 가지고 직렬화를 시킬 수 있도록 도움을 주도록 설정

* 이외 약간의 변경사항 적용

- DB schema 변경 (block) 에 따른 쿼리문 수정

- 차단 유무를 확인하는 작업은 읽기 작업이므로 slave에서 읽어올 수 있도록
  변경

- 추천글을 남길 때 친구 여부를 확인하는 과정에서 자기 자신이 시도하는
  경우도 있으므로 예외 메시지에 `자기 자신` 문구도 추가

msugo1 committed:
Merge branch 'develop' into feature/36
